xarray/coding/variables.py

```diff
@@ -458,9 +485,15 @@ def decode(self, variable: Variable, name: T_Name = None):
 
         if encoded_fill_values:
             # special case DateTime to properly handle NaT
+            # we need to check if time-like will be decoded or not
+            # in further processing
             dtype: np.typing.DTypeLike
             decoded_fill_value: Any
-            if _is_time_like(attrs.get("units")) and data.dtype.kind in "iu":
+            is_time_like = _is_time_like(attrs.get("units"))
+            if (
+                (is_time_like == "datetime" and self.decode_times)
+                or (is_time_like == "timedelta" and self.decode_timedelta)
+            ) and data.dtype.kind in "iu":
```

Comment:
If I understand correctly, unpacked data will always be of floating point dtype, so should we always decode the fill value to NaN in the case of packed times?  Otherwise I believe `np.iinfo(np.int64).min` could lose its meaning during the unpacking step.

Reply:
You are correct, this is a flaw in the current decoding. I'll restructure decoding part of `CFMaskCoder` accordingly.